### ISSUE TITLE
[Bug]: malformed url params break up calypso

### ISSUE THREAD

**Issue by kozer** [Automattic/wp-calypso#69794]
### The problem
When there is a malformed URL parameter, calypso breaks.

Look here for an example:

Normal URL:
![normal-url](https://user-images.githubusercontent.com/497103/200026799-252110fc-d078-472e-aa39-9d6c7901eb76.png)

Malformed URL:
![malformed-url](https://user-images.githubusercontent.com/497103/200026842-ad317521-6eec-447b-ab54-a4ec5884ea24.png)

This is not an easy fix as we use different ways to parse URL parameters across the project.

The issue is happening as `decodeURIComponent` and `decodeURI` functions throw an error if they found a malformed URL.

There are several places where we use `qs` library for example, or other approaches to parse URL params. This has the same effect (probably because they use the same js functions underneath)

Try for example to navigate [here](https://wordpress.com/start/domains?ref=calypso-selector&source=my-home&siteSlug=%E0%A4%A) . 
(in the above screen we use `qs` library to parse query parameters)

However wherever we use the `getQueryArgs`, either from `@wordpress/url` library, or the one inside `client/lib/query-args` we can easily fix that issue, by changing the function implementation.

### Steps to reproduce

1. Visit `http://calypso.localhost:3000/start/domains?ref=calypso-selector&source=my-home&siteSlug=<your-domain>`. Site works fine.
2. Visit `http://calypso.localhost:3000/start/domains?ref=calypso-selector&source=my-home&siteSlug=%E0%A4%A`. Site breaks.

### What you expected to happen

Malformed URL params should not break the site

### What actually happened

Calypso breaks

### Reproducibility

Consistent

### Severity

All

### Available workarounds?

Yes, difficult to implement

### Workaround details

- Update `README.md` and state that the preferred way to parse parameters is by using `getQueryArgs` from `@wordpress/url` library
- Change the places where we use something different (like `qs` library for example)
- Update the `@wordpress/url` package after accepting [this](https://github.com/WordPress/gutenberg/pull/45561) pr. This will solve the problem in places that `getQueryArgs` is used.

**Comment by danielbachhuber** [comment]
@kozer Once the next version of `@wordpress/url` is [tagged](https://github.com/WordPress/gutenberg/pull/45561#issuecomment-1382045649), I think you should be able to update it here to finally resolve the issue.

**Comment by danielbachhuber** [comment]
@kozer Was `@wordpress/url` updated in Calypso?

**Comment by danielbachhuber** [comment]
> Was `@wordpress/url` updated in Calypso?

@kozer Can you look into this when you have a moment?

**Comment by kozer** [comment]
@danielbachhuber , it seems that the url fix was added in [15.1.0 RC1](https://github.com/WordPress/gutenberg/releases/tag/v15.1.0-rc.1) and the `@wordpress/url` is on version `3.27.0`.

Unfortunately, calypso's version is still in `3.10.0`.

Should I update it indivudually? Or the dependencies are updated altogether by the Calypso team?

**Comment by danielbachhuber** [comment]
> Should I update it indivudually? Or the dependencies are updated altogether by the Calypso team?

@kozer What did you find when looking at past dependency updates?

**Comment by kozer** [comment]
> > Should I update it indivudually? Or the dependencies are updated altogether by the Calypso team?
> 
> @kozer What did you find when looking at past dependency updates?

If I understood correctly, there is a `renovate` bot that does the dependency update? Is this correct?

**Comment by danielbachhuber** [comment]
> If I understood correctly, there is a `renovate` bot that does the dependency update? Is this correct?

@kozer Hm... where did you find that?

Looking at the [history for yarn.lock](https://github.com/Automattic/wp-calypso/commits/trunk/yarn.lock), it doesn't appear dependencies are updated with any consistency:

<img width="1278" alt="image" src="https://user-images.githubusercontent.com/36432/216710208-647571fa-9b5f-4dd0-9e02-a680fb6f3c49.png">

@Automattic/team-calypso-frameworks Do you have any preference on how `@wordpress/url` updates are brought in?

**Comment by noahtallen** [comment]
> Or the dependencies are updated altogether by the Calypso team?

Normally we do handle the `@wordpress/*` updates. The packages are so intertwined that a one-off update will normally not work immediately. When updating them all together, there are typically several things to fix to get it working. As a result, it often gets out of date until there's a reason to update it (like this case!)

Renovate is useful for just getting the updates into a branch before the extra work that needs done, but it's currently disabled after a security incident last year. But it's also not very useful here because of the extra work involved in updating WordPress deps

In other words, I can look at this next week.

**Comment by kozer** [comment]
>Renovate is useful for just getting the updates into a branch before the extra work that needs done, but it's currently disabled after a security incident last year. But it's also not very useful here because of the extra work involved in updating WordPress deps

>In other words, I can look at this next week.

Thanks @noahtallen ! And for the info as well!

**Comment by tyxla** [comment]
Thanks for looking into it @noahtallen!

**Comment by noahtallen** [comment]
One challenge with this upgrade is that it may require us to update to react 18 (or be in violation of peer dependency requests), so I could see it being a bigger task

**Comment by danielbachhuber** [comment]
It looks like `@wordpress/url` made it to v3.23.0 with https://github.com/Automattic/wp-calypso/pull/73890

<img width="666" alt="image" src="https://user-images.githubusercontent.com/36432/224851241-6944f22f-8ce6-40cc-b225-5ee476a7b22d.png">

**Comment by noahtallen** [comment]
With a quick test (https://wordpress.com/start/domains?ref=calypso-selector&source=my-home&siteSlug=%E0%A4%A), it's still a WSOD apparently, but it doesn't show the old error screen.

So there could be an issue with how that page handles sites that don't exist on the client side

**Comment by tyxla** [comment]
I agree with @noahtallen. This appears to be because of the poor way we handle the site request failures in signup:

https://github.com/Automattic/wp-calypso/blob/2eb5007b9669f61fa72afbdf2faf59ddd239f740/client/signup/controller.js#L347-L350

Something as simple as this, for example:

```
diff --git a/client/signup/controller.js b/client/signup/controller.js
index 5a4d6b21a0..fd2f43c11b 100644
--- a/client/signup/controller.js
+++ b/client/signup/controller.js
@@ -3,6 +3,7 @@ import { isEmpty } from 'lodash';
 import page from 'page';
 import { createElement } from 'react';
 import store from 'store';
+import { notFound } from 'calypso/controller';
 import { recordPageView } from 'calypso/lib/analytics/page-view';
 import { login } from 'calypso/lib/paths';
 import { sectionify } from 'calypso/lib/route';
@@ -345,7 +346,7 @@ export default {
                        // Fetch the site by siteIdOrSlug and then try to select again
                        dispatch( requestSite( siteIdOrSlug ) )
                                .catch( () => {
-                                       next();
+                                       notFound( context, next );
                                        return null;
                                } )
                                .then( () => {
```

Will display the following error page when testing with @noahtallen's quick test:

![Screenshot 2023-03-14 at 12 41 12](https://user-images.githubusercontent.com/8436925/224975793-b1dd8928-62f9-42c1-a065-182835dcfe83.png)

Of course, it might make more sense for this message to be more personalized, but I'll leave that to whoever is working on the fix. Happy to help with reviews or further ideas, of course.

**Comment by kozer** [comment]
> With a quick test (https://wordpress.com/start/domains?ref=calypso-selector&source=my-home&siteSlug=%E0%A4%A), it's still a WSOD apparently, but it doesn't show the old error screen.
So there could be an issue with how that page handles sites that don't exist on the client side

You are right on that. changing the `@wordpress/url` is not solving the problem completely, in all cases. Only in those places that `@wordpress/url` is used to parse the parameters. 

Your change solves the problem in that specific case, but not everywhere. 

We need to gradually remove externals and use  `@worpdress/url` everywhere.

However, as a step moving forward, I created a pr, introducing your changes, and I will close this issue.
